x/mongo/driver/topology/connection.go

```diff
@@ -108,7 +109,24 @@ func (c *connection) connect(ctx context.Context) {
 	}
 	defer close(c.connectDone)
 
+	var cancelFn context.CancelFunc
+
+	c.connectContextMutex.Lock()
 	ctx, c.cancelConnectContext = context.WithCancel(ctx)
+	c.connectContextMutex.Unlock()
+
+	defer func() {
```

Comment:
I think there's a nested `defer` because of my previous comment to defer mutex unlocks rather than doing them manually, but the nested `defer` is actually making things harder to read. Can this be simplified to basically be the same code as the `closeConnectContext` function?